platforms/cpuinfo_linux.go

```diff
@@ -0,0 +1,137 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package platforms
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"os"
+	"runtime"
+	"strings"
+
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/log"
+	"golang.org/x/sys/unix"
+)
+
+// getCPUInfo retrieves field value of "info" in file /proc/cpuinfo
+// For Linux, the kernel has already detected the ABI, ISA and Features.
+// So we don't need to access the ARM registers to detect platform information
+// by ourselves. We can just parse these information from /proc/cpuinfo
+func getCPUInfo(pattern string) (info string, err error) {
+
+	cpuinfo, err := os.Open("/proc/cpuinfo")
+	if err != nil {
+		return "", err
+	}
+	defer cpuinfo.Close()
+
+	// Start to Parse the Cpuinfo line by line. For SMP SoC, we parse
+	// the first core is enough.
+	scanner := bufio.NewScanner(cpuinfo)
+	for scanner.Scan() {
+		newline := scanner.Text()
+		list := strings.Split(newline, ":")
+
+		if len(list) > 1 && strings.EqualFold(strings.TrimSpace(list[0]), pattern) {
+			return strings.TrimSpace(list[1]), nil
+		}
+	}
+
+	// Check whether the scanner encountered errors
+	err = scanner.Err()
+	if err != nil {
+		return "", err
+	}
+
+	return "", fmt.Errorf("getCPUInfo for pattern: %s: %w", pattern, errdefs.ErrNotFound)
+}
+
+// getCPUArchitecture gets CPU Architecture from RawSyscall
+func getCPUArchitecture() (string, error) {
+
+	var name unix.Utsname
+	err := unix.Uname(&name)
+	if err != nil {
+		return "", err
+	}
+	machine := string(name.Machine[:bytes.IndexByte(name.Machine[:], 0)])
+	// aarch64 is alternatively used as armv8, as it doesn't follow the same
+	//name convention as armv7, armv6, we return it directly and it will be normalized later
+	if machine == "aarch64" {
+		return machine, nil
+	}
+	length := len(machine)
+	// remove last 1 char l or b
+	variant := machine[:length-1]
+	if len(variant) <= 4 {
+		return "", fmt.Errorf("getCPUArchitecture for unix.Uname for machine field invalid %s", machine)
+	}
+	// remove first 4 char armv
+	variant = variant[3:]
+	return variant, nil
+}
+
+// getCPUVariant returns cpu variant
+func getCPUVariant() string {
+
+	variant, err := getCPUInfo("Cpu architecture")
+	if err != nil {
+		//If we can't find variant from /proc/cpuinfo, let's try uname -m.
+		//This issue was seen when server and client running in different architectures
```

Comment:
I have synced-up with @fangn2 and this comment doesn't capture his intention well. Apparently QEMU's user-mode emulation doesn't emulate /proc/cpuinfo, but wraps uname syscall.

Reply:
Yes, I deprecated the function, and use uname syscall directly to get CPU architecture.